data_prep/graph_dataset.py

```diff
@@ -1,72 +1,155 @@
             tf_idf_words (list): List of words according to the tf.idf matrix.
             pmi_scores (dict): Dictionary of word pairs and corresponding PMI scores.
+            ntoi (dict): Dictionary mapping from nodes to indices.
         Returns:
             edge_index (Tensor): List of edges.
             edge_attr (Tensor): List of edge weights.
         """
-        raise NotImplementedError
+        edge_index = []
+        edge_attr = []
+
+        # Document-word edges
+        for d_ind, doc in enumerate(tf_idf):
+            tf_idf_inds = doc.indices
+            for tf_idf_ind in tf_idf_inds:
+                # Convert index from tf.idf to index in ntoi
+                word = tf_idf_words[tf_idf_ind]
+                w_ind = ntoi[word]
 
-    def generate_masks(self, train_num, val_num, test_num):
+                edge_index.append([d_ind, w_ind])
+                edge_index.append([w_ind, d_ind])
+                edge_attr.append(tf_idf[d_ind, tf_idf_ind])
+                edge_attr.append(tf_idf[d_ind, tf_idf_ind])
+
+        # Word-word edges
+        for (word_i, word_j), score in pmi_scores.items():
+            w_i_ind = ntoi[word_i]
+            w_j_ind = ntoi[word_j]
+            edge_index.append([w_i_ind, w_j_ind])
+            edge_index.append([w_j_ind, w_i_ind])
+            edge_attr.append(score)
+            edge_attr.append(score)
+
+        edge_index = torch.tensor(edge_index).t().contiguous()
+        edge_attr = torch.tensor(edge_attr).float()
+        return edge_index, edge_attr
+
+    def _generate_masks(self, train_num, val_num, test_num, all_num):
         """
         Generates masking for the different splits in the dataset.
         Args:
             train_num (int): Number of training documents.
             val_num (int): Number of validation documents.
             test_num (int): Number of test documents.
+            all_num (int): Number of all nodes, including words
         Returns:
             train_mask (Tensor): Training mask as boolean tensor.
             val_mask (Tensor): Validation mask as boolean tensor.
             test_mask (Tensor): Test mask as boolean tensor.
         """
-        raise NotImplementedError
+        train_mask = torch.zeros(all_num, device=self._device)
+        train_mask[:train_num] = 1
+
+        val_mask = torch.zeros(all_num, device=self._device)
+        val_mask[train_num:train_num + val_num] = 1
+
+        # Mask all non-test docs
+        test_mask = torch.zeros(all_num, device=self._device)
+        test_mask[val_num + train_num:val_num + train_num + test_num] = 1
+
+        return train_mask.bool(), val_mask.bool(), test_mask.bool()
 
     @abc.abstractmethod
-    def preprocess(self):
+    def _preprocess(self):
         """
         Preprocesses the corpus.
         """
         raise NotImplementedError
 
     @abc.abstractmethod
-    def generate_features(self):
+    def _generate_features(self):
```

Comment:
If we are not calling this method from the init method in this class, it does not need to be here in the superclass I guess.

Reply:
I thought this would similar to the num_classes thing, where we enforce a unified interface

Reply:
If it's not called here. Should it still be a private method?

Reply:
It is called inside the init of the subclasses so it's still not to be used from "outside". It's not part of the interface